### PR TITLE
add support for flat both inner and outer dims

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Hi,

I'm trying to flat both the inner and outer dims and only keeps the middle ones but I didn't find any existing method in tensorflow that support this.  So I implement one. Hope this is helpful and can be merged into tensorflow.

A brief changelog:

1. add new method `flat_inner_outer_dims`
2. change private method `ComputeFlatInnerDims` and `ComputeFlatOuterDims` to static member
3. add tests for `flat_inner_outer_dims`

### CODE REVIEW

File: tensorflow/core/framework/tensor.cc [hunk 1]
```diff
@@ -902,43 +902,31 @@ void Tensor::FillDescription(TensorDescription* description) const {
...
-  int64 prod_out_dims = 1;
-  for (int64 out_dim = num_out_dims - 1; out_dim > 0; --out_dim) {
-    const int64 in_dim = out_dim + (dims() - num_out_dims);
-    out_dims[out_dim] = (in_dim >= dims() || in_dim < 0) ? 1 : dim_size(in_dim);
-    prod_out_dims *= out_dims[out_dim];
-  }
-  if (prod_out_dims != 0) {
-    out_dims[0] = num_elements / prod_out_dims;
-  } else {
-    out_dims[0] = 0;
```
Comment (zasdfgbnm) [inline tensorflow/core/framework/tensor.cc]: Actually, I think there is a bug in the old code that is not discovered by the test case. The old code uses `prod_out_dims` to keep track of the product of the last `num_out_dims` elements and compute the first element of `out_dim` as `num_elements / prod_out_dims`. In the case of `prod_out_dims == 0` it don't make any sense to set the first dimension as `0`.  For example, if you have a tensor with shape `{5,3,0,0,0}` and you run a `flat_inner_dims<?,4>`, you should get a tensor with shape `{15,0,0,0}`. But the way it was computed will gives a `{0,0,0,0}`

File: tensorflow/core/framework/tensor.cc [hunk 2]
```diff
@@ -902,43 +902,31 @@ void Tensor::FillDescription(TensorDescription* description) const {
...
-  const int64 num_elements = NumElements();
-  int64 prod_out_dims = 1;
-  for (int64 out_dim = 0; out_dim < num_out_dims - 1; ++out_dim) {
-    out_dims[out_dim] = out_dim >= dims() ? 1 : dim_size(out_dim);
-    prod_out_dims *= out_dims[out_dim];
-  }
-  if (prod_out_dims != 0) {
-    out_dims[num_out_dims - 1] = num_elements / prod_out_dims;
-  } else {
-    out_dims[num_out_dims - 1] = 0;
```
Comment (zasdfgbnm) [inline tensorflow/core/framework/tensor.cc]: Same thing as I mentioned above.

**Comment by keveman** [comment]
@tensorflow-jenkins test this please.

File: tensorflow/core/framework/tensor.cc [hunk 3]
```diff
@@ -902,43 +902,31 @@ void Tensor::FillDescription(TensorDescription* description) const {
...
-  if (prod_out_dims != 0) {
-    out_dims[0] = num_elements / prod_out_dims;
-  } else {
-    out_dims[0] = 0;
+  int64 offset = orig.size() - num_out_dims;
+  for (int64 out_dim = num_out_dims - 1; out_dim >= 0; --out_dim) {
+    const int64 in_dim = out_dim + offset;
+    out_dims[out_dim] = in_dim < 0 ? 1 : orig[in_dim];
   }
+  for (int64 in_dim = 0; in_dim < offset; ++in_dim)
```
Comment (rmlarsen) [inline tensorflow/core/framework/tensor.cc]: Please use braces around the statement in the loop.

File: tensorflow/core/framework/tensor.cc [hunk 4]
```diff
@@ -902,43 +902,31 @@ void Tensor::FillDescription(TensorDescription* description) const {
...
-  for (int64 out_dim = 0; out_dim < num_out_dims - 1; ++out_dim) {
-    out_dims[out_dim] = out_dim >= dims() ? 1 : dim_size(out_dim);
-    prod_out_dims *= out_dims[out_dim];
-  }
-  if (prod_out_dims != 0) {
-    out_dims[num_out_dims - 1] = num_elements / prod_out_dims;
-  } else {
-    out_dims[num_out_dims - 1] = 0;
-  }
+  for (int64 out_dim = 0; out_dim <= num_out_dims - 1; ++out_dim)
```
Comment (rmlarsen) [inline tensorflow/core/framework/tensor.cc]: braces

File: tensorflow/core/framework/tensor.cc [hunk 5]
```diff
@@ -902,43 +902,31 @@ void Tensor::FillDescription(TensorDescription* description) const {
...
+    out_dims[out_dim] = out_dim >= orig.size() ? 1 : orig[out_dim];
+  for (int64 in_dim = num_out_dims; in_dim < orig.size(); ++in_dim)
```
Comment (rmlarsen) [inline tensorflow/core/framework/tensor.cc]: braces

File: tensorflow/core/framework/tensor.h [hunk 6]
```diff
@@ -638,22 +652,34 @@ typename TTypes<T>::ConstScalar Tensor::scalar() const {
...
 
 template <typename T, size_t NDIMS>
 typename TTypes<T, NDIMS>::Tensor Tensor::flat_outer_dims() {
-  return shaped<T, NDIMS>(ComputeFlatOuterDims(NDIMS));
+  return shaped<T, NDIMS>(ComputeFlatOuterDims(shape_.dim_sizes(), NDIMS));
+}
+
+template <typename T, size_t NDIMS>
+typename TTypes<T, NDIMS>::Tensor Tensor::flat_inner_outer_dims(int64 begin) {
+  gtl::InlinedVector<int64, 4> o = ComputeFlatOuterDims(shape_.dim_sizes(), begin+NDIMS);
```
Comment (rmlarsen) [inline tensorflow/core/framework/tensor.h]: Please eschew abbreviation and use a more meaningful name than 'o'.

Comment (rmlarsen) [inline tensorflow/core/framework/tensor.h]: space around '+'

File: tensorflow/core/framework/tensor.h [hunk 7]
```diff
@@ -638,22 +652,34 @@ typename TTypes<T>::ConstScalar Tensor::scalar() const {
...
 }
 
 template <typename T, size_t NDIMS>
 typename TTypes<T, NDIMS>::ConstTensor Tensor::flat_outer_dims() const {
-  return shaped<T, NDIMS>(ComputeFlatOuterDims(NDIMS));
+  return shaped<T, NDIMS>(ComputeFlatOuterDims(shape_.dim_sizes(), NDIMS));
+}
+
+template <typename T, size_t NDIMS>
+typename TTypes<T, NDIMS>::Tensor Tensor::flat_inner_outer_dims(int64 begin) const {
```
Comment (rmlarsen) [inline tensorflow/core/framework/tensor.h]: ditto: rename 'o', space around '+'

Comment (zasdfgbnm) [inline tensorflow/core/framework/tensor.h]: fixed

**Comment by zasdfgbnm** [comment]
I will be traveling for the following week and will come back to continue on this after that.

**Comment by keveman** [comment]
@tensorflow-jenkins test this please

File: tensorflow/core/framework/tensor.h [hunk 8]
```diff
@@ -638,22 +651,36 @@ typename TTypes<T>::ConstScalar Tensor::scalar() const {
...
 
 template <typename T, size_t NDIMS>
 typename TTypes<T, NDIMS>::Tensor Tensor::flat_outer_dims() {
-  return shaped<T, NDIMS>(ComputeFlatOuterDims(NDIMS));
+  return shaped<T, NDIMS>(ComputeFlatOuterDims(shape_.dim_sizes(), NDIMS));
+}
+
+template <typename T, size_t NDIMS>
+typename TTypes<T, NDIMS>::Tensor Tensor::flat_inner_outer_dims(int64 begin) {
+  gtl::ArraySlice<int64> flat_outer = ComputeFlatOuterDims(
```
Comment (rmlarsen) [inline tensorflow/core/framework/tensor.h]: I think this needs to be an InlinedVector too.

**Comment by vrv** [comment]
(Marking as stalled since the author is out for a few more days).

**Comment by zasdfgbnm** [comment]
everything should works well now

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
@zasdfgbnm Thanks for the contribution!

**Comment by zasdfgbnm** [comment]
`ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job.`
What does this mean?

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by rmlarsen** [comment]
@zasdfgbnm This appear to be an internal error in the test system, unrelated to your change. I think this is ready to merge.

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by rmlarsen** [comment]
@yifeif @gunan do you know why the cmake keeps failing with "ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job."

**Comment by yifeif** [comment]
Updated the job, let's try again. @tensorflow-jenkins test this please.

**Comment by rmlarsen** [comment]
@yifeif it looks like another flaky test failed this time. I think it is safe to merge this.

**Pull request merged**